packages/block-library/src/categories/index.php

```diff
@@ -16,13 +16,19 @@ function render_block_core_categories( $attributes ) {
 	static $block_id = 0;
 	$block_id++;
 
+	// WP_Term_Query `number` parameter uses 0 to represent "all".
+	// https://developer.wordpress.org/reference/classes/wp_term_query/__construct/#parameters
+	$term_query_show_all = 0;
+
 	$args = array(
 		'echo'         => false,
 		'hierarchical' => ! empty( $attributes['showHierarchy'] ),
-		'orderby'      => 'name',
+		'orderby'      => $attributes['orderBy'],
+		'order'        => $attributes['order'],
 		'show_count'   => ! empty( $attributes['showPostCounts'] ),
 		'title_li'     => '',
 		'hide_empty'   => empty( $attributes['showEmpty'] ),
+		'number'       => $attributes['termsToShow'] === -1 ? $term_query_show_all : $attributes['termsToShow'],
 	);
 	if ( ! empty( $attributes['showOnlyTopLevel'] ) && $attributes['showOnlyTopLevel'] ) {
 		$args['parent'] = 0;
```

Comment:
Changing this to `$args['child_of'] = 0;` seems to fix the issue with the `number` limit on the front end, but I'm not sure if this is the correct fix.